lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -25,95 +26,114 @@ using namespace circt;
-                ArrayAttr::get(send.getContext(), {xmrDefOp});
-            opsToRemove.push_back(send);
+            Value xmrDef = send.getBase();
+            // Get an InnerRefAttr to the xmrDef op. If the operation does not
+            // take any InnerSym (like firrtl.add, firrtl.or etc) then create a
+            // NodeOp to add the InnerSym.
+            if (!xmrDef.isa<BlockArgument>()) {
+              Operation *xmrDefOp = xmrDef.getDefiningOp();
+              if (!isa<InnerSymbolOpInterface>(xmrDefOp)) {
+                // Add a node, for non-innerSym ops. Otherwise the sym will be
+                // dropped after LowerToHW.
+                ImplicitLocOpBuilder b(xmrDefOp->getLoc(), xmrDefOp);
+                b.setInsertionPointAfter(xmrDefOp);
+                StringRef opName;
+                auto nameKind = NameKindEnum::DroppableName;
+                if (auto name = xmrDefOp->getAttrOfType<StringAttr>("name")) {
+                  opName = name.getValue();
+                  nameKind = NameKindEnum::InterestingName;
+                }
+                xmrDef = b.create<NodeOp>(xmrDef.getType(), xmrDef, opName,
+                                          nameKind);
+              }
+            }
+            // Create a new entry for this RefSendOp. The path is currently
+            // local.
+            addReachingSendsEntry(send.getResult(), getInnerRefTo(xmrDef));
+            markForRemoval(send);
             return success();
           })
           .Case<InstanceOp>([&](auto inst) { return handleInstanceOp(inst); })
           .Case<FConnectLike>([&](FConnectLike connect) {
             // Ignore BaseType.
             if (!connect.getSrc().getType().isa<RefType>())
               return success();
-            // Get the dataflow value into the src.
-            if (auto remoteOpPath = getRemoteRefSend(connect.getSrc()))
-              reachingRefSendAt[connect.getDest()] = remoteOpPath;
-            else
-              return failure();
-            opsToRemove.push_back(connect);
+            markForRemoval(connect);
+            // Merge the dataflow classes of destination into the source of the
+            // Connect. This handles two cases:
+            // 1. If the dataflow at the source is known, then the
+            // destination is also inferred. By merging the dataflow class of
+            // destination with source, every value reachable from the
+            // destination automatically infers a reaching RefSend.
+            // 2. If dataflow at source is unkown, then just record that both
+            // source and destination will have the same dataflow information.
+            // Later in the pass when the reaching RefSend is inferred at the
+            // leader of the dataflowClass, then we automatically infer the
+            // dataflow at this connect and every value reachable from the
+            // destination.
+            dataFlowClasses.unionSets(
```

Comment:
Oh wow, thanks for using EC's!  This looks great!

Maintaining a map from leaders to values is a bit tricky, as the leader for a given class is not stable across union operations.  For this reason, this and the other `unionSets` call may need to be careful and update/merge/drop mappings for the old leader (in `dataflowAt`).  The new leader should be one of the leaders of the unioned sets.  InferResets has some handling for similar scenario, IIRC.  Maybe one day we can wrap this in a ECMap or something to handle this for us...

Reply:
I checked the union implementation, seems like it is deterministic, its always L1's leader, can you confirm by checking this ?
https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/EquivalenceClasses.h#L242
```
 member_iterator unionSets(member_iterator L1, member_iterator L2) {
    assert(L1 != member_end() && L2 != member_end() && "Illegal inputs!");
    if (L1 == L2) return L1;   // Unifying the same two sets, noop.

    // Otherwise, this is a real union operation.  Set the end of the L1 list to
    // point to the L2 leader node.
    const ECValue &L1LV = *L1.Node, &L2LV = *L2.Node;
    L1LV.getEndOfList()->setNext(&L2LV);

    // Update L1LV's end of list pointer.
    L1LV.Leader = L2LV.getEndOfList();

    // Clear L2's leader flag:
    L2LV.Next = L2LV.getNext();

    // L2's leader is now L1.
    L2LV.Leader = &L1LV;
    return L1;
  }
```

Reply:
Wow, that seems right to me! Great to know, thanks! :+1: 

I'm not sure if this is something we can assume to be true, but I'm satisfied.  Didn't want to be chasing down some bug in a large design that triggered this :grin:.

Does this mean `dataflowAt` doesn't need handling for merging the old leader's data?

Reply:
I had the same concern earlier, but at least going through the implementation, seems like the leader is always consistent, and there is no "re-shuffling" of leaders ! Let's stick with this assumption for now.
> Does this mean `dataflowAt` doesn't need handling for merging the old leader's data?

So, the dataflow direction ensures that the merge always implies there is no entry in `dataflowAt` , for the old class leader being merged. So, there is no updated to be done.